include/drivers/video.h

```diff
@@ -0,0 +1,556 @@
+	video_api_set_ctrl_t set_ctrl;
+	video_api_set_ctrl_t get_ctrl;
+	video_api_register_event_callback_t register_event_callback;
+};
+
+/**
+ * @brief Set video format.
+ *
+ * Configure video device with a specific format.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param fmt Pointer to a video format struct.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_set_format(struct device *dev,
+				   enum video_endpoint_id ep,
+				   struct video_format *fmt)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	__ASSERT(api->set_format, "set_format must be implemented by driver");
+
+	return api->set_format(dev, ep, fmt);
+}
+
+/**
+ * @brief Get video format.
+ *
+ * Get video device current video format.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param fmt Pointer to video format struct.
+ *
+ * @retval pointer to video format
+ */
+static inline int video_get_format(struct device *dev,
+				   enum video_endpoint_id ep,
+				   struct video_format *fmt)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	__ASSERT(api->get_format, "get_format must be implemented by driver");
+
+	return api->get_format(dev, ep, fmt);
+}
+
+/**
+ * @brief Enqueue a video buffer.
+ *
+ * Enqueue an empty (capturing) or filled (output) video buffer in the driver’s
+ * endpoint incoming queue.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param buf Pointer to the video buffer.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_enqueue(struct device *dev, enum video_endpoint_id ep,
+				struct video_buffer *buf)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->enqueue == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->enqueue(dev, ep, buf);
+}
+
+/**
+ * @brief Dequeue a video buffer.
+ *
+ * Dequeue a filled (capturing) or displayed (output) buffer from the driver’s
+ * enpoint outgoing queue.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param buf Pointer a video buffer pointer.
+ * @param timeout Timeout in milliseconds.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_dequeue(struct device *dev, enum video_endpoint_id ep,
+				struct video_buffer **buf,
+				u32_t timeout)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->dequeue == NULL) {
```

Comment:
Same comment as above, it seems like this and enqueue should be required instead of optional

Reply:
Well, there is some configuration implying non-bufferable video devices, such devices are 'control' only (set/get_format, set/get-ctrls...). These devices are generally slaves and controlled by a master video device (and not directly by the user). This is the case for example on nxp mimxrt1064-evk board, there are two video devices, the SoC video interface (NXP CSI), and the sensor (e.g. MT9M114) controlled via I2C, but outputting useful data to the parallel interface (CSI). In that case CSI device is the main interface the user deal with and CSI driver control internally the 'slave' which can be MT9M114 sensor or any other one.

 USER
      |
 [NXP-CSI] <-> [MT9M114-I2C-SENSOR]